src/helpers/pivot/spreadsheet_pivot/spreadsheet_pivot.ts

```diff
@@ -0,0 +1,388 @@
+        this.coreDefinition,
+        this.fields,
+        this.getters
+      );
+      this.table = undefined;
+      this.dataEntries = [];
+      const range = this._definition.range;
+      if (this.isValid() && range) {
+        this.dataEntries = this.extractDataEntriesFromRange(range);
+      }
+      this.needsReevaluation = false;
+    }
+  }
+
+  get isInvalidRange() {
+    return !!this.invalidRangeError;
+  }
+
+  get invalidRangeMessage() {
+    return this.invalidRangeError?.message;
+  }
+
+  get definition() {
+    if (!this._definition) {
+      throw new Error("Pivot not loaded yet");
+    }
+    return this._definition;
+  }
+
+  isValid(): boolean {
+    if (this.invalidRangeError || !this._definition) {
+      return false;
+    }
+    for (const measure of this.definition.measures) {
+      if (!measure.isValid) {
+        return false;
+      }
+    }
+    for (const dimension of this.definition.columns.concat(this.definition.rows)) {
+      if (!dimension.isValid) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  assertIsValid({ throwOnError }: { throwOnError: boolean }): FPayload | undefined {
+    if (!this.isValid()) {
+      if (throwOnError) {
+        if (this.invalidRangeError) {
+          throw this.invalidRangeError;
+        } else {
+          throw new EvaluationError(_t("At least one measure and/or dimension is not correct."));
+        }
+      }
+      return {
+        value: CellErrorType.GenericError,
+        message:
+          this.invalidRangeError?.message ??
+          _t("At least one measure and/or dimension is not correct."),
+      };
+    }
+    return undefined;
+  }
+
+  markAsDirtyForEvaluation(): void {
+    this.needsReevaluation = true;
+  }
+
+  getMeasure(name: string): PivotMeasure {
+    return this.definition.getMeasure(name);
+  }
+
+  getPivotHeaderValueAndFormat(domainStr: StringDomainArgs): FPayload {
+    const domain = toDomainArgs(domainStr);
+    const lastNode = domain.at(-1);
+    if (!lastNode) {
+      return { value: _t("Total") };
+    }
+    if (lastNode.field === "measure") {
+      return { value: this.getMeasure(lastNode.value).displayName };
+    }
+    const dimension = this.getDimension(lastNode.field);
+    const cells = this.filterDataEntriesFromDomain(this.dataEntries, domain);
+    const finalCell = cells[0]?.[dimension.nameWithGranularity];
+
+    if (!finalCell) {
+      return { value: "" };
+    }
+    if (finalCell.value === null) {
+      return { value: _t("(Undefined)") };
+    }
+    if (dimension.type === "date") {
+      if (dimension.granularity === "day") {
+        return {
+          value: toNumber(finalCell.value, this.getters.getLocale()),
+          format: this.getters.getLocale().dateFormat,
+        };
+      }
+      if (dimension.granularity === "month_number") {
+        return {
+          value: MONTHS[toNumber(finalCell.value, this.getters.getLocale())].toString(),
+        };
+      }
```

Comment:
this should probably be defined somewhere else in a map of { "granularity" : valueFormatStuff() }

Reply:
I think it's too specific to create a map with that. It will only have two entries, and the two will have a different object (one with value and format, one with only value). As we only have two cases, I think it's better like this.
What do you think ?

Reply:
we can still refactor later, when we have more cases :)